### PR TITLE
Added notification service

### PR DESCRIPTION
**Pull request by hirdesharma**
## What ?
NotificationService consists of Kafka consumers to handle subscription and user registration events, along with the necessary configuration and exception handling.

- KafkaConsumerConfig: Configures Kafka consumer properties and sets up the KafkaListenerContainerFactory.
- KafkaSubscriptionListeners: Listens to subscription events and sends confirmation emails using EmailSenderService.
- KafkaUserRegistrationListener: Listens to user registration events and sends registration confirmation emails using EmailSenderService.
- NotificationServiceException: A custom exception for handling errors within the notification service.

## Why?

- This implementation is aimed at handling asynchronous events related to user subscriptions and registrations, ensuring that users receive appropriate email notifications based on their interactions with the system. By using Kafka, the service can scale effectively and handle high volumes of events.

## Impact

- Kafka Integration: Introduces Kafka configuration and message listening, which may affect the performance and resource utilization of the service.
- Email Notifications: Sends emails based on received Kafka messages, which could impact the volume of outgoing emails.
- Exception Handling: Adds custom exception handling to manage errors gracefully and provide informative error messages.

## Testing

- Listeners: Tests have been created for both KafkaSubscriptionListeners and KafkaUserRegistrationListener to ensure that the message processing and email sending functionalities work correctly.
- Service: The EmailSenderService is tested to ensure that it correctly sends emails as expected.

## Test Cases
[ Test Cases, HLD, LLD ](https://docs.google.com/document/d/1KFvsDH-w52EGL86THDYFq8MDVWd-aPSrN1xDwZYPdms/edit?usp=sharing)
## Design Patterns Followed

- Dependency Injection: Utilized Spring's dependency injection for managing service and configuration beans.
- Exception Handling: Implemented custom exceptions to handle and report errors effectively.

## Checklist

- [x]  I have self-reviewed the PR.
- [x]  I am aware of the impact of the changes on services.
- [x]  I have verified that all relevant tests are included and have passed.

### CODE REVIEW

File: notification-service/src/main/java/com/example/notification/listener/KafkaSubscriptionListeners.java [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+    try {
+      final JsonNode jsonNode = objectMapper.readTree(data);
+      final String userId = jsonNode.get("id").asText();
+      final String email = jsonNode.get("email").asText();
+
+      System.out.println("Processed subscription for user: " + userId);
+      final String body = "Congratulations you { userId : " + userId + "} have been subscribed to "
+          + "our "
+          + "subscription service";
+      emailSenderService.sendEmail(email, body);
```
Comment (sankethkini) [inline notification-service/src/main/java/com/example/notification/listener/KafkaSubscriptionListeners.java]: form the email body inside emailSenderService

File: notification-service/src/main/java/com/example/notification/listener/KafkaSubscriptionListeners.java [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+public class KafkaSubscriptionListeners {
+
+  private final EmailSenderService emailSenderService;
+  private final ObjectMapper objectMapper = new ObjectMapper();
+
+  @KafkaListener(topics = "${spring.kafka.topic.subscription-events}",
+                 groupId = "${spring.kafka.group.subscription-notification}")
+  void listener(final String data) {
+    try {
+      final JsonNode jsonNode = objectMapper.readTree(data);
```
Comment (sankethkini) [inline notification-service/src/main/java/com/example/notification/listener/KafkaSubscriptionListeners.java:35]: its ok for now but its always cleaner approach to parse json string to java class and then use it. U can add validation also

File: notification-service/src/test/java/com/example/notification/listener/KafkaSubscriptionListenersTest.java [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+
+  @BeforeEach
+  void setUp() {
+    MockitoAnnotations.openMocks(this);
+  }
+
+  @Test
+  void testListener_Success() throws Exception {
+    String data = "{\"id\":\"12345\",\"email\":\"test@example.com\"}";
+    JsonNode jsonNode = mock(JsonNode.class);
```
Comment (sankethkini) [inline notification-service/src/test/java/com/example/notification/listener/KafkaSubscriptionListenersTest.java]: why mocking JsonNode? u can actually use objectMapper and then parse the data string right?

File: notification-service/src/test/java/com/example/notification/listener/KafkaSubscriptionListenersTest.java [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+import com.example.notification.service.EmailSenderService;
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.mockito.InjectMocks;
+import org.mockito.Mock;
+import org.mockito.MockitoAnnotations;
+
+class KafkaSubscriptionListenersTest {
```
Comment (sankethkini) [inline notification-service/src/test/java/com/example/notification/listener/KafkaSubscriptionListenersTest.java:17]: test negative scenarios as well

File: notification-service/src/main/java/com/example/notification/listener/KafkaSubscriptionListeners.java (hunk 1 above)
Comment (hirdesharma) [reply notification-service/src/main/java/com/example/notification/listener/KafkaSubscriptionListeners.java]: in that case we will have to create two different functions emailSenderService with everything same  except body

is it fine to do so?

Comment (hirdesharma) [reply notification-service/src/main/java/com/example/notification/listener/KafkaSubscriptionListeners.java]: done

File: notification-service/src/main/java/com/example/notification/listener/KafkaSubscriptionListeners.java (hunk 2 above)
Comment (hirdesharma) [reply notification-service/src/main/java/com/example/notification/listener/KafkaSubscriptionListeners.java:35]: okay i will remember

File: notification-service/src/test/java/com/example/notification/listener/KafkaSubscriptionListenersTest.java (hunk 3 above)
Comment (hirdesharma) [reply notification-service/src/test/java/com/example/notification/listener/KafkaSubscriptionListenersTest.java]: Yes mocking JsonNode isn't needed used objectMapper in its place and parsing the data string

File: notification-service/src/test/java/com/example/notification/listener/KafkaSubscriptionListenersTest.java (hunk 4 above)
Comment (hirdesharma) [reply notification-service/src/test/java/com/example/notification/listener/KafkaSubscriptionListenersTest.java:17]: added some negative test case scenarios

**Pull request merged**